### DISCUSSION TITLE
Reflections from a plugin dev

### DISCUSSION THREAD

**Discussion by woutersamaey** [btcpayserver/btcpayserver#4991] [Ideas / Feature Requests]
After coding 2 plugins now, I thought it be a good idea to collect my thoughts on plugin development and make a list of what I found to be missing in the whole experience.

- Better plugin dev doc. The current one contains mistakes and is too confusing. Will try to address this myself when I get a bit of time.
- No live reload at runtime. I know this is probably a tough one, but even basic template/css/js/image reloading would be nice.
- No easy way to do EF Migrations - but coding by hand is fine. If the docs would just say so, it would've saved me a lot of time! Not sure I really need EF migrations to be honest. Worst case, I reset the DB and start fresh.
- No unit tests. It was suggested to me to put tests in the original test project, but this way it's not part of the plugin repo + I cannot test classes from the plugin. I'm left with only "generic" stuff I can test.
- Huge lack of hooks and filters. Looking at WordPress, the "framework" has a logical list if hooks and filters that are present throughout the app. These are well thought of and can cater to any situation.
- No way to modify existing pages, like grids or forms. These should IMHO be replaced with grid and form models where we can use code to "add column", "add cell renderer", "add form field", "add mass action" etc... Magento has this extensively and is great. I think we can copy those ideas.
- No way to run code before/after data is saved or actions are taken. Often times you want to intervene/prevent a save, or change some data last-minute, but this is not possible. EF has a solution for this, but we need EF > 6.2 for that.
- No swagger combining of my plugin's extra API endpoint. Having 1 swagger containing everything is still awesome, but plugins are not part of it yet. 
- Logging could be better if we were to have a log file per plugin. To debug users, asking them for the plugin log file would be convenient.

**Comment by pavlenex** [comment]
Thanks for collecting your thoguhts, I think this is really helpful, since on of our focus should def be to improve developers experiences when building plugins. cc @NicolasDorier maybe this is interesting feedback for you.
